memflow/src/plugins/os.rs

```diff
@@ -0,0 +1,160 @@
+{
+}
+impl<T: Process + Clone, K: 'static + Clone + for<'a> OSInner<'a, IntoProcessType = T>> PluginOS<T>
+    for K
+{
+}
+
+pub fn create_with_logging<P: 'static + Process + Clone, T: PluginOS<P>>(
+    args: ReprCStr,
+    conn: ConnectorInstance,
+    log_level: i32,
+    out: &mut MUOSInstance,
+    create_fn: impl Fn(&Args, ConnectorInstance, log::Level) -> Result<T>,
+) -> i32 {
+    super::util::create_with_logging(args, log_level, out, move |a, l| {
+        create_fn(&a, conn, l).map(OSInstance::new)
+    })
+}
+
+pub fn create_without_logging<P: 'static + Process + Clone, T: PluginOS<P>>(
+    args: ReprCStr,
+    conn: ConnectorInstance,
+    out: &mut MUOSInstance,
+    create_fn: impl Fn(&Args, ConnectorInstance) -> Result<T>,
+) -> i32 {
+    super::util::create_without_logging(args, out, |a| create_fn(&a, conn).map(OSInstance::new))
+}
+
+#[repr(C)]
+pub struct OSLayerDescriptor {
+    /// The connector inventory api version for when the connector was built.
+    /// This has to be set to `MEMFLOW_PLUGIN_VERSION` of memflow.
+    ///
+    /// If the versions mismatch the inventory will refuse to load.
+    pub os_version: i32,
+
+    /// The name of the connector.
+    /// This name will be used when loading a connector from a connector inventory.
+    pub name: &'static str,
+
+    /// Create instance of the OS
+    pub create: extern "C" fn(ReprCStr, COption<ConnectorInstance>, i32, &mut MUOSInstance) -> i32,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct OSLayerFunctionTable {
+    /// The vtable for object creation and cloning
+    pub base: OpaqueBaseTable,
+    /// The vtable for all os functions
+    pub os: OpaqueOSFunctionTable,
+    /// The vtable for all physical memory access if available
+    pub phys: Option<&'static OpaquePhysicalMemoryFunctionTable>,
+    /// The vtable for all virtual memory access if available
+    pub virt: Option<&'static OpaqueVirtualMemoryFunctionTable>,
+}
+
+impl OSLayerFunctionTable {
+    pub fn new<P: 'static + Process + Clone, T: PluginOS<P>>() -> Self {
+        OSLayerFunctionTable {
+            base: GenericBaseTable::<T>::default().into_opaque(),
+            os: OSFunctionTable::<P, T>::default().into_opaque(),
+            phys: None,
+            virt: None,
+        }
+    }
+}
+
+pub struct LoadableOS {
+    descriptor: OSLayerDescriptor,
+}
+
+impl Loadable for LoadableOS {
+    type Instance = OSInstance;
+    type InputArg = Option<ConnectorInstance>;
+
+    fn ident(&self) -> &str {
+        self.descriptor.name
+    }
+
+    fn load(library: &CArc<Library>, path: impl AsRef<Path>) -> Result<LibInstance<Self>> {
+        let descriptor = unsafe {
+            library
+                .as_ref()
+                .get::<*mut OSLayerDescriptor>(b"MEMFLOW_OS\0")
+                .map_err(|_| Error::Connector("OS descriptor not found"))?
+                .read()
+        };
+
+        if descriptor.os_version != MEMFLOW_PLUGIN_VERSION {
```

Comment:
We should use goblin here to not require us to load the library before actually checking if it's a memflow plugin. This would also prevent us from potential security issues by loading random libraries into the process.

This should also be a separate tracking issue.